### PR TITLE
Vision: Add gRPC support for image properties.

### PR DESCRIPTION
**Pull request by daspecster**
~~Based from #2922~~

Adds gRPC support for image properties detection.

### CODE REVIEW

File: system_tests/vision.py [hunk 1]
```diff
@@ -478,13 +478,16 @@ def tearDown(self):
...
-        self.assertIsInstance(color.red, int)
-        self.assertIsInstance(color.green, int)
-        self.assertIsInstance(color.blue, int)
+        self.assertIsInstance(color.red, (int, float))
+        self.assertIsInstance(color.green, (int, float))
+        self.assertIsInstance(color.blue, (int, float))
+        if not isinstance(color.alpha, float):
+            self.assertIsInstance(color.alpha.value, float)
+        else:
+            self.assertIsInstance(color.alpha, float)
```
Comment (tseaver) [inline system_tests/vision.py]: Grrr:  I wish the back-ends would just be consistent.  In what world are RGB values floats?

Comment (daspecster) [inline system_tests/vision.py]: Right?...and why is it that just `alpha` is a special `FloatValue` instead of a native `float`.

File: vision/google/cloud/vision/color.py [hunk 2]
```diff
@@ -26,20 +26,38 @@ def __init__(self, colors):
...
+        :param image_properties: Protobuf response from Vision API with image
+                                 properties data.
+
+        :rtype: list of
+                :class:`~google.cloud.vision.color.ImagePropertiesAnnotation`
+        :returns: List of ``ImagePropertiesAnnotation``.
+        """
+        colors = getattr(image_properties.dominant_colors, 'colors', ())
+        if len(colors) == 0:
+            return ()
```
Comment (daspecster) [inline vision/google/cloud/vision/color.py]: If we want a tuple returned instead of an empty list I can do this.

File: system_tests/vision.py (hunk 1 above)
Comment (dhermes) [inline system_tests/vision.py]: @daspecster Which backend is returned which? ISTM you should be cleaning this before / while creating a color instance?

File: vision/google/cloud/vision/color.py (hunk 2 above)
Comment (dhermes) [inline vision/google/cloud/vision/color.py]: This breaks "principle" of least surprise. You should either return `None` or an instance of `cls`

File: vision/google/cloud/vision/color.py [hunk 3]
```diff
@@ -26,20 +26,38 @@ def __init__(self, colors):
...
+        return [cls([ColorInformation.from_pb(color) for color in colors])]
```
Comment (dhermes) [inline vision/google/cloud/vision/color.py]: Why a list? Should it just be an instance of `cls`?

File: vision/google/cloud/vision/color.py [hunk 4]
```diff
@@ -93,6 +111,22 @@ def from_api_repr(cls, response):
 
         return cls(red, green, blue, alpha)
 
+    @classmethod
+    def from_pb(cls, color):
+        """Factory: construct a ``Color`` from a protobuf response.
+
+        :type color: :class:`~google.type.color_pb2`
```
Comment (dhermes) [inline vision/google/cloud/vision/color.py]: This is a module, not a class (not that it will show up via Intersphinx either way)

File: vision/google/cloud/vision/color.py [hunk 5]
```diff
@@ -93,6 +111,22 @@ def from_api_repr(cls, response):
...
+        :param color: ``Color`` from API Response.
+
+        :rtype: :class:`~google.cloud.vision.color.Color`
+        :returns: Instance of :class:`~google.cloud.vision.color.Color`.
+        """
+        red = getattr(color, 'red', 0.0)
+        green = getattr(color, 'green', 0.0)
+        blue = getattr(color, 'blue', 0.0)
+        alpha = getattr(color.alpha, 'value', 0.0)
```
Comment (dhermes) [inline vision/google/cloud/vision/color.py]: Why are all of these `getattrs` necessary?

```python
>>> import google.type.color_pb2
>>> C1 = google.type.color_pb2.Color()
>>> C2 = google.type.color_pb2.Color(red=0.5)
>>> import google.protobuf.wrappers_pb2
>>> alpha = google.protobuf.wrappers_pb2.FloatValue(value=0.5)
>>> C3 = google.type.color_pb2.Color(
...     red=1.0, green=0.25, blue=1.0, alpha=alpha)
>>> (C1.red, C2.red, C3.red)
(0.0, 0.5, 1.0)
>>> (C1.green, C2.green, C3.green)
(0.0, 0.0, 0.25)
>>> (C1.blue, C2.blue, C3.blue)
(0.0, 0.0, 1.0)
>>> (C1.alpha.value, C2.alpha.value, C3.alpha.value)
(0.0, 0.0, 0.5)
```

File: vision/google/cloud/vision/color.py [hunk 6]
```diff
@@ -149,19 +183,34 @@ def __init__(self, color, score, pixel_fraction):
...
+        :type color_information: dict
+        :param color_information: Color data with extra meta information.
 
         :rtype: :class:`~google.cloud.vision.color.ColorInformation`
         :returns: Instance of ``ColorInformation``.
         """
-        color = Color.from_api_repr(response.get('color'))
-        score = response.get('score')
-        pixel_fraction = response.get('pixelFraction')
+        color = Color.from_api_repr(color_information.get('color'))
```
Comment (dhermes) [inline vision/google/cloud/vision/color.py]: What happens if `Color.from_api_repr` gets `None`? Don't you want to fall back to a dict?

File: vision/google/cloud/vision/color.py [hunk 7]
```diff
@@ -149,19 +183,34 @@ def __init__(self, color, score, pixel_fraction):
...
+        score = color_information.get('score')
+        pixel_fraction = color_information.get('pixelFraction')
```
Comment (dhermes) [inline vision/google/cloud/vision/color.py]: Is `None` the right default value for these two?

File: system_tests/vision.py (hunk 1 above)
Comment (daspecster) [inline system_tests/vision.py]: Just the protobuf/gRPC backend. I can't find it now, but there was a little comment block in either the protobuf definition or the generated python code about why it's there. Basically to tell if it's been set or not...if I remember right.

Comment (dhermes) [inline system_tests/vision.py]: But isn't `color` here one of your instances?

Comment (daspecster) [inline system_tests/vision.py]: Ha! Yes it is. #fixed in next commit.

File: vision/google/cloud/vision/color.py (hunk 7 above)
Comment (daspecster) [inline vision/google/cloud/vision/color.py]: I *think* so because `None` would mean that they were not set.

File: vision/google/cloud/vision/color.py (hunk 3 above)
Comment (daspecster) [inline vision/google/cloud/vision/color.py]: I can change it, but I would want to change the HTTP version to match at the same time. Should that be a different PR?

File: vision/google/cloud/vision/color.py [hunk 8]
```diff
@@ -26,20 +26,37 @@ def __init__(self, colors):
...
+                :class:`~google.cloud.vision.color.ImagePropertiesAnnotation`
+        :returns: List of ``ImagePropertiesAnnotation``.
         """
-        raw_colors = response.get('dominantColors', {}).get('colors', ())
-        colors = [ColorInformation.from_api_repr(color)
-                  for color in raw_colors]
-        return cls(colors)
+        colors = getattr(image_properties.dominant_colors, 'colors', ())
+        if len(colors) > 0:
+            return [cls([ColorInformation.from_pb(color) for color in colors])]
```
Comment (dhermes) [inline vision/google/cloud/vision/color.py]: > I can change it, but I would want to change the HTTP version to match at the same time. Should that be a different PR?

I am looking at the HTTP version on line 41-42:

```python
        return cls([ColorInformation.from_api_repr(color)
                    for color in colors])
```

It returns an instance, not a list.

Comment (daspecster) [inline vision/google/cloud/vision/color.py]: It turns into a list later on. Good catch though! I updated it in https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2937/commits/6e39ce0cbf9bf090d47ae515a3f66d1158dc6913.

File: system_tests/vision.py [hunk 9]
```diff
@@ -474,13 +474,13 @@ def tearDown(self):
             value.delete()
 
     def _assert_color(self, color):
-        self.assertIsInstance(color.red, int)
-        self.assertIsInstance(color.green, int)
-        self.assertIsInstance(color.blue, int)
+        self.assertIsInstance(color.red, (int, float))
+        self.assertIsInstance(color.green, (int, float))
+        self.assertIsInstance(color.blue, (int, float))
```
Comment (dhermes) [inline system_tests/vision.py]: Is `int` still possible?

File: system_tests/vision.py [hunk 10]
```diff
@@ -493,19 +493,14 @@ def _assert_properties(self, image_property):
...
-        self._pb_not_implemented_skip(
-            'gRPC not implemented for image properties detection.')
         client = Config.CLIENT
         with open(FACE_FILE, 'rb') as image_file:
             image = client.image(content=image_file.read())
         properties = image.detect_properties()
-        self.assertEqual(len(properties), 1)
-        image_property = properties[0]
-        self._assert_properties(image_property)
+        self.assertEqual(len(properties[0]))
```
Comment (dhermes) [inline system_tests/vision.py]: Did you mean to change the actual output type? This is a "breaking API change" that is not at all related to "Add gRPC support for image properties."

Comment (daspecster) [inline system_tests/vision.py]: I don't think it's a breaking change.
It currently returns a [`list`](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/309fa8d6a9268875eb9d7f141342f583b75c5d23/system_tests/vision.py#L503).

Is that what you mean?

Comment (daspecster) [inline system_tests/vision.py]: Never mind I see what you mean. My typo.

File: system_tests/vision.py (hunk 9 above)
Comment (daspecster) [inline system_tests/vision.py]: Yeah HTTP version returns ints. I could cast it in `Color` though. WDYT?

Comment (dhermes) [inline system_tests/vision.py]: The HTTP version return ints just an artifact of JSON encoding stripping extra data?

```node
> JSON.stringify({a: 1.0})
"{"a":1}"
```

to

```python
>>> import json
>>> json.loads('{"a": 1}')
{u'a': 1}
>>> # vs.
>>> json.loads('{"a": 1.0}')
{u'a': 1.0}
```

Should we cast them to floats?

File: system_tests/vision.py [hunk 11]
```diff
@@ -493,19 +493,14 @@ def _assert_properties(self, image_property):
...
-        self._pb_not_implemented_skip(
-            'gRPC not implemented for image properties detection.')
         client = Config.CLIENT
         with open(FACE_FILE, 'rb') as image_file:
             image = client.image(content=image_file.read())
         properties = image.detect_properties()
         self.assertEqual(len(properties), 1)
-        image_property = properties[0]
-        self._assert_properties(image_property)
+        self._assert_properties(properties[0])
```
Comment (dhermes) [inline system_tests/vision.py]: Not that it matters, but this doesn't need to be in the diff (i.e. the changed code did the exact same thing)

File: vision/unit_tests/test_color.py [hunk 12]
```diff
@@ -45,3 +71,36 @@ def test_missing_rgb_values(self):
...
         self.assertEqual(colors.alpha, 0.0)
+
+
+class TestImagePropertiesAnnotation(unittest.TestCase):
+    @staticmethod
+    def _get_target_class():
+        from google.cloud.vision.color import ImagePropertiesAnnotation
+        return ImagePropertiesAnnotation
+
+    def test_color_annotation_from_pb(self):
```
Comment (dhermes) [inline vision/unit_tests/test_color.py]: You need a test that covers the case when `None` is returned in `from_pb`

File: vision/google/cloud/vision/annotations.py [hunk 13]
```diff
@@ -89,8 +89,11 @@ def from_api_repr(cls, response):
         annotations = {}
         for feature_type, annotation in response.items():
             curr_feature = annotations.setdefault(_KEY_MAP[feature_type], [])
-            curr_feature.extend(
-                _entity_from_response_type(feature_type, annotation))
+            entity = _entity_from_response_type(feature_type, annotation)
+            if isinstance(entity, list):
+                curr_feature.extend(entity)
+            else:
+                annotations[_KEY_MAP[feature_type]] = entity
```
Comment (dhermes) [inline vision/google/cloud/vision/annotations.py]: Why was this necessary? I would say:

- Just keep returning a list in `_entity_from_response_type`
- Why did we suddenly grow this feature where `annotations` gains an entry?

Comment (daspecster) [inline vision/google/cloud/vision/annotations.py]: It didn't gain an entry, it's just handling for setting 'properties' to not be a list like the others so it matches the protobuf version.

Comment (dhermes) [inline vision/google/cloud/vision/annotations.py]: OK. We should try to clean it up later?

Comment (daspecster) [inline vision/google/cloud/vision/annotations.py]: Yeah, I'm thinking of handling it more like how I'm doing the pb versions with the helper methods that do the conversions.

WDYT?

File: system_tests/vision.py (hunk 9 above)
Comment (daspecster) [inline system_tests/vision.py]: Cast to floats https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2937/commits/a39be5fe5a9be88de9cf501856589d1965a3e409

**Comment by daspecster** [comment]
I opend https://github.com/googleapis/googleapis/issues/242 for the RGB values.

File: vision/unit_tests/test_color.py (hunk 12 above)
Comment (daspecster) [inline vision/unit_tests/test_color.py]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2937/commits/8809f2ffa72a3f664ae46418e59f5e12940b8998

Comment (dhermes) [inline vision/unit_tests/test_color.py]: @daspecster That test doesn't call `ImagePropertiesAnnotation.from_pb`

Comment (daspecster) [inline vision/unit_tests/test_color.py]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2937/commits/9d40edbb4daa9ff21451234e08042284e65275fe

Comment (daspecster) [inline vision/unit_tests/test_color.py]: I updated the names as well to be a little more clear.

File: vision/google/cloud/vision/annotations.py (hunk 13 above)
Comment (dhermes) [inline vision/google/cloud/vision/annotations.py]: OK, re-reading this, it seems it could benefit from a helper function and then just be:

```python
annotations = {
    _KEY_MAP[feature_type]: _toms_helper(feature_type, annotation),
    for feature_type, annotation in six.iteritems(response)
}
return cls(**annotations)
```

I **WAS** worried about key collision, but after examining the structure it's clear there wouldn't be one (provided the values in `_KEY_MAP` are unique).

I was also thrown off by the fact that you grab

```python
curr_feature = annotations.setdefault(_KEY_MAP[feature_type], [])
```

independent of the value of `isinstance(entity, list)` (you only need it in the `True` branch).

Comment (daspecster) [inline vision/google/cloud/vision/annotations.py]: Unless you want me to add to this PR, I'll follow up after this and #2938.

Comment (dhermes) [inline vision/google/cloud/vision/annotations.py]: Please add to this PR

Comment (daspecster) [inline vision/google/cloud/vision/annotations.py]: Updated in https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2937/commits/bd7fd857439c66e2d01c966c6d0c45d0c6ad99dd.

**Comment by dhermes** [comment]
LGTM
